File: rag/domain/followup.py
"""Rewriting a dependent follow-up into a self-contained query.

Within a Conversation a follow-up turn often leans on what was just asked - a
referential pronoun ("what is the punishment for *it*?") or an elliptical
connector ("and *cheating*?"). On its own such a query carries no statutory
content to ground on, so it would be refused. This module rewrites it into a
standalone query by folding in the bounded recent Conversation context, before
the query enters retrieval. A query that is already self-contained is returned
unchanged, so non-follow-up turns are unaffected.

The rewrite is deliberately deterministic and dependency-free, matching the
offline retrieval and generation seams. Production swaps a Claude-backed
(claude-opus) rewriter behind :func:`rewrite_followup`; the surface contract -
recent context plus a follow-up in, one standalone query out - stays the same.
"""
from __future__ import annotations

import re
from typing import Sequence

# Words that mark a query as leaning on earlier context: back-referential
# pronouns and the elliptical connectors that open a continued question. When a
# query carries one of these it is treated as a follow-up and resolved against
# the recent context rather than answered on its own.
_FOLLOWUP_MARKERS = frozenset(
    {
        "it",
        "its",
        "that",
        "this",
        "them",
        "they",
        "those",
        "these",
        "one",
        "ones",
        "also",
        "else",
    }
)

_WORD_RE = re.compile(r"[a-z]+")


def is_followup(query: str) -> bool:
    """Whether ``query`` reads as a follow-up that depends on prior context."""
    words = _WORD_RE.findall(query.lower())
    return any(word in _FOLLOWUP_MARKERS for word in words)


def rewrite_followup(query: str, recent_context: Sequence[str]) -> str:
    """Rewrite a dependent follow-up into a standalone query using recent context.

    ``recent_context`` is the bounded list of recent standalone queries in the
    Conversation, oldest first. A follow-up is made self-contained by prefixing
    that context so retrieval sees the subject it refers back to; a self-contained
    query (or one with no context yet) is returned unchanged.
    """
    if not recent_context or not is_followup(query):
        return query
    return " ".join([*recent_context, query])

File: web/convex/lib/followup.ts
// Rewriting a dependent follow-up into a self-contained query, ported from
// rag/domain/followup.py. A referential pronoun ("what is the punishment for
// *it*?") or elliptical connector carries no statutory content of its own and
// would be refused; folding in the bounded recent Conversation context first
// lets retrieval see the subject it refers back to. Deliberately
// deterministic and dependency-free.

// Words that mark a query as leaning on earlier context.
const FOLLOWUP_MARKERS = new Set([
  "it", "its", "that", "this", "them", "they", "those", "these",
  "one", "ones", "also", "else",
]);

const WORD_RE = /[a-z]+/g;

// Whether `query` reads as a follow-up that depends on prior context.
export function isFollowup(query: string): boolean {
  const words = query.toLowerCase().match(WORD_RE) ?? [];
  return words.some((word) => FOLLOWUP_MARKERS.has(word));
}

// Rewrite a dependent follow-up into a standalone query using recent context
// (the bounded list of recent standalone queries, oldest first). A
// self-contained query, or one with no context yet, is returned unchanged.
export function rewriteFollowup(
  query: string,
  recentContext: string[],
): string {
  if (recentContext.length === 0 || !isFollowup(query)) {
    return query;
  }
  return [...recentContext, query].join(" ");
}

// How many recent standalone turns seed follow-up rewriting, everywhere a
// Conversation's memory is rebuilt (mirrors Conversation._CONTEXT_TURNS).
export const CONTEXT_TURNS = 4;
